modules/security-aai/src/main/java/org/opencastproject/security/aai/DynamicLoginHandler.java

```diff
@@ -0,0 +1,432 @@
+    JpaOrganization organization = fromOrganization(securityService.getOrganization());
+
+    // Compile the list of roles
+    Set<JpaRole> roles = extractRoles(id, request);
+
+    // Create the user reference
+    JpaUserReference userReference = new JpaUserReference(id, name, email, MECH_SHIBBOLETH, loginDate, organization,
+            roles);
+
+    logger.debug("Shibboleth user '{}' logged in for the first time", id);
+    userReferenceProvider.addUserReference(userReference, MECH_SHIBBOLETH);
+  }
+
+  /**
+   * Handle an existing user login.
+   *
+   * @param id
+   *          The identity of the user, ideally the Shibboleth persistent unique identifier
+   * @param request
+   *          The request, for accessing any other Shibboleth variables
+   */
+  @Override
+  public void existingUserLogin(String id, HttpServletRequest request) {
+    Organization organization = securityService.getOrganization();
+
+    // Load the user reference
+    JpaUserReference userReference = userReferenceProvider.findUserReference(id, organization.getId());
+    if (userReference == null) {
+      //Triggers creation of user reference
+      //Possible problem: if there is user (not reference) with that id, we will get conflicts
+      throw new UsernameNotFoundException("User reference '" + id + "' was not found");
+    }
+
+    // Update the reference
+    userReference.setName(extractName(request));
+    userReference.setEmail(extractEmail(request));
+    userReference.setLastLogin(new Date());
+    Set<JpaRole> roles = extractRoles(id, request);
+    userReference.setRoles(roles);
+
+    logger.debug("Shibboleth user '{}' logged in", id);
+    userReferenceProvider.updateUserReference(userReference);
+  }
+
+  /**
+   * Sets the security service.
+   *
+   * @param securityService
+   *          the security service
+   */
+  public void setSecurityService(SecurityService securityService) {
+    this.securityService = securityService;
+  }
+
+  /**
+   * Sets the user reference provider.
+   *
+   * @param userReferenceProvider
+   *          the user reference provider
+   */
+  public void setUserReferenceProvider(UserReferenceProvider userReferenceProvider) {
+    this.userReferenceProvider = userReferenceProvider;
+  }
+
+  /**
+   * Extracts the name from the request.
+   *
+   * @param request
+   *          the request
+   * @return the name
+   */
+  private String extractName(HttpServletRequest request) {
+    String displayName = extractDisplayName(request);
+    if (displayName != null && !"".equals(displayName)) {
+      return displayName;
+    }
+    List<String> givenNames = attributeMapper.getMappedAttributes(request, "givenName");
+    String givenName = "";
+    if (givenNames.size() > 0) {
+      String givenNameValue = givenNames.get(0);
+      givenName = StringUtils.isBlank(givenNameValue) ? ""
+              : new String(givenNameValue.getBytes(StandardCharsets.ISO_8859_1),
+                      StandardCharsets.UTF_8);
```

Comment:
And if `givenNameValue` is blank wouldn't `new String(givenNameValue.getBytes(StandardCharsets.ISO_8859_1),                      StandardCharsets.UTF_8);` still be a blank string?

Reply:
Wenn givenNameValue is blank, givenName is Blank (""), otherwise not. Same code in ConfigurableLoginHandler and for the other attributes. Should be OK.

The whole building of the displayName is only a fallback since it's also possible to configure the building of the displayName in SPeL configuration...

Reply:
It is not necesary

Reply:
Code is unnecesatry